test/ranges.coffee

```diff
@@ -116,3 +116,99 @@ test "#1012 slices with arguments object", ->
 
 test "#1409: creating large ranges outside of a function body", ->
   CoffeeScript.eval '[0..100]'
+
+test "#2047: Infinite loop possible when `for` loop with `range` uses variables", ->
+  up = 1
+  down = -1
+  a = 1
+  b = 5
+
+  testRange = (arg) ->
+    [from, to, step, expectedResult] = arg
+    r = (x for x in [from..to] by step)
+    arrayEq r, expectedResult
+
+  testData = [
+    [1, 5, 1, [1..5]]
+    [1, 5, -1, [1]]
+    [1, 5, up, [1..5]]
+    [1, 5, down, [1]]
+
+    [a, 5, 1, [1..5]]
+    [a, 5, -1, [1]]
+    [a, 5, up, [1..5]]
+    [a, 5, down, [1]]
+
+    [1, b, 1, [1..5]]
+    [1, b, -1, [1]]
+    [1, b, up, [1..5]]
+    [1, b, down, [1]]
+
+    [a, b, 1, [1..5]]
+    [a, b, -1, [1]]
+    [a, b, up, [1..5]]
+    [a, b, down, [1]]
+
+    [5, 1, 1, [5]]
+    [5, 1, -1, [5..1]]
+    [5, 1, up, [5]]
+    [5, 1, down,  [5..1]]
+
+    [5, a, 1, [5]]
+    [5, a, -1, [5..1]]
+    [5, a, up, [5]]
+    [5, a, down, [5..1]]
+
+    [b, 1, 1, [5]]
+    [b, 1, -1, [5..1]]
+    [b, 1, up, [5]]
+    [b, 1, down, [5..1]]
+
+    [b, a, 1, [5]]
+    [b, a, -1, [5..1]]
+    [b, a, up, [5]]
+    [b, a, down, [5..1]]
+  ]
+
+  testRange d for d in testData
+
+test "#2047: from, to and step as variables", ->
+  up = 1
+  down = -1
+  a = 1
+  b = 5
+
+  r = (x for x in [a..b] by up)
+  arrayEq r, [1..5]
+
+  r = (x for x in [a..b] by down)
+  arrayEq r, [1]
+
+  r = (x for x in [b..a] by up)
+  arrayEq r, [5]
+
+  r = (x for x in [b..a] by down)
+  arrayEq r, [5..1]
+
+  a = 1
+  b = -1
+  step = 0
+  r = (x for x in [b..a] by step)
+  arrayEq r, []
+
+  a = "aaa"
+  b = "bbb"
+  step = "3"
+  r = (x for x in [b..a] by step)
+  arrayEq r, ['bbb']
+
+  step = "ccc"
+  r = (x for x in [b..a] by step)
+  arrayEq r, ['bbb']
+
+  a = 2
+  b = 1
+  step = "0.5000000"
```

Comment:
Should we really be supporting step values or variables that aren’t integers? The `for` loops in these cases becomes code like this:

```js
for (x = i = ref = b, ref1 = a, ref2 = step; ref2 !== 0 && (ref <= ref1 ? ref <= i && i <= ref1 : ref >= i && i >= ref1); x = i += ref2) {
```

with the relevant part being the end, `x = i += ref2`. So basically our iterator is getting incremented by the value of `step`, as it should be, but I’m pretty sure in that `i +=` part it was assumed that the incrementing value would be an integer.

Get rid of the `r =` in these tests and you’ll see why. Because of the `r =`, there’s an implicit `results` array that `x` is getting pushed to, so we never need to reference an index on the array we’re looping through. However, without that:

```coffee
arr = [1..3]
step = '0.1'
for x in arr by step
  console.log x
```
```js
var arr, i, len, ref, step, x;

arr = [1, 2, 3];

step = '0.1';

ref = step;
for ((ref > 0 ? (i = 0, len = arr.length) : i = arr.length - 1); ref > 0 ? i < len : i >= 0; i += ref) {
  x = arr[i];
  console.log(x);
}
```

That `x = arr[i]` line doesn’t work if `i` isn’t an integer.

Reply:
I see what you mean. But, the real difference is whether we use `Identifier` or `Range` in the `for` loop.

```coffeescript
arr = [1..2]
step = 0.5
for x in arr by step
  console.log x

###
ref = step;
for ((ref > 0 ? (i = 0, len = arr.length) : i = arr.length - 1); ref > 0 ? i < len : i >= 0; i += ref) {
  x = arr[i];
  console.log(x);
}

Ouput:
1
undefined
2
undefined
###

a = 1
b = 2
step = 0.5
for x in [a..b] by step
  console.log x

###
for (x = i = ref = a, ref1 = b, ref2 = step; ref2 !== 0 && (ref <= ref1 ? ref <= i && i <= ref1 : ref >= i && i >= ref1); x = i += ref2) {
  console.log(x);
}

Ouput:
1
1.5
2
###
```

The former will check `arr.length` and the [latter](http://coffeescript.org/#try:a%20%3D%201%0Ab%20%3D%202%0Astep%20%3D%200.5%0Afor%20x%20in%20%5Ba..b%5D%20by%20step%0A%20%20console.log%20x%0A%0A%20%20) the upper bound of the range, e.g. `b`.
With `Range` I would expect this result. For example, Ruby and Python support float values for step in ranges.

I guess the question is if we should throw an error when the `step` isn't integer when not used with `Range` (e.g. `for x in arr by 0.5`).

Reply:
> I guess the question is if we should throw an error when the `step` isn’t integer when not used with `Range`

But how would we know that, if `step` is a variable?

Reply:
It looks that `step` is never cast into a number. In case `step` isn't a number, at least first item from the `array` or `range` will pass.

```coffeescript
arr = [10..1]
step = "a"
for x in arr by step
  console.log x

### 
output:
1
###

for x in [5..1] by step
  console.log x

### 
output:
5
###
```

```coffeescript
arr = [10..1]
step = "-0.5"
for x in arr by step
  console.log x

### 
output:
1
###

for x in [5..1] by step
  console.log x

### 
output:
5
###
```

```coffeescript
arr = [10..1]
step = -0.5
for x in arr by step
  console.log x

### 
output:
1
undefined
2
undefined
3
undefined
4
undefined
5
undefined
6
undefined
7
undefined
8
undefined
9
undefined
10
###

for x in [5..1] by step
  console.log x

### 
output:
5
4.5
4
3.5
3
2.5
2
1.5
1
###
```

Reply:
So in other words, the iterator that gets created is `0` by default, and after the first loop `+= step` gets applied to it, which breaks it?

Reply:
So I guess dealing with the case where the step variable isn’t numeric needn’t be part of the scope of this PR, since it no longer is casting the step variable to a number. The current compiler also has the issue where a non-numeric step variable will allow the first iteration of a loop over an array to be run, but none thereafter. I don’t know if that’s a bug, necessarily, but it surprised me.

I think we should rewrite these tests that use string-based steps, though, as they’re misleading; when I see `step = "0.500000"` I’m assuming you’re testing that that string value will be cast as a float. Looking at this test:

```coffee
a = 2
b = 1
step = "0.5000000"
r = (x for x in [b..a] by step)
arrayEq r, [1]
```

Why would I expect `r` to equal `[1]`? What are you really testing here? A quick read of this test implies that `r` should be `[ 1, 1.5, 2 ]`, assuming that `by` is supposed to cast the string or if I didn’t notice that the step was a string rather than a float. The same goes for the other “`step` is a string” tests. It might help if some of the more complicated ones have comments explaining what they’re intending to test, and/or why the output is as expected.

Reply:
So for the purposes of wrapping up this PR, can we take out the “string `step`“ tests? And we can discuss that in a future issue/PR if you’d like.

Reply:
I removed "string" tests.